dataflow/src/main/java/org/checkerframework/dataflow/analysis/FlowExpressions.java

```diff
@@ -1108,6 +1116,206 @@ public String toString() {
+            return right;
+        }
+
+        @Override
+        public boolean containsOfClass(Class<? extends Receiver> clazz) {
+            if (getClass() == clazz) {
+                return true;
+            }
+            return left.containsOfClass(clazz) || right.containsOfClass(clazz);
+        }
+
+        @Override
+        public boolean isUnassignableByOtherCode() {
+            return left.isUnassignableByOtherCode() && right.isUnassignableByOtherCode();
+        }
+
+        @Override
+        public boolean isUnmodifiableByOtherCode() {
+            return left.isUnmodifiableByOtherCode() && right.isUnmodifiableByOtherCode();
+        }
+
+        @Override
+        public boolean syntacticEquals(Receiver other) {
+            if (!(other instanceof BinaryOperation)) {
+                return false;
+            }
+            BinaryOperation biOp = (BinaryOperation) other;
+            if (!(operationKind == biOp.getOperationKind())) {
+                return false;
+            }
+            return left.equals(biOp.left) && right.equals(biOp.right);
+        }
+
+        @Override
+        public boolean containsModifiableAliasOf(Store<?> store, Receiver other) {
+            return left.containsModifiableAliasOf(store, other)
+                    || right.containsModifiableAliasOf(store, other);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(operationKind, left, right);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (!(other instanceof BinaryOperation)) {
+                return false;
+            }
+            BinaryOperation biOp = (BinaryOperation) other;
+            if (!(operationKind == biOp.getOperationKind())) {
+                return false;
+            }
+            if (isCommutative()) {
+                return (left.equals(biOp.left) && right.equals(biOp.right))
+                        || (left.equals(biOp.right) && right.equals(biOp.left));
+            }
+            return left.equals(biOp.left) && right.equals(biOp.right);
+        }
+
+        /**
+         * Check if the binary operation is commutative. e.g. x + y == y + x.
+         *
+         * @return true if the binary operation is commutative.
+         */
+        private boolean isCommutative() {
+            switch (operationKind) {
+                case PLUS:
+                case MULTIPLY:
+                case AND:
+                case OR:
+                case XOR:
+                case EQUAL_TO:
+                case NOT_EQUAL_TO:
+                case CONDITIONAL_AND:
+                case CONDITIONAL_OR:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder result = new StringBuilder();
+            result.append(left.toString());
+
+            switch (operationKind) {
```

Comment:
Instead of doing this yourself, can you maybe use `com.sun.tools.javac.tree.Pretty`?
https://github.com/openjdk/jdk/blob/master/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java#L1216 and you would need the JCTree.Tag in addition to the Kind...
Instantiating Pretty is easy: https://github.com/typetools/checker-framework/blob/master/framework/src/main/java/org/checkerframework/common/util/debug/TreePrinter.java#L46